File: src/main/java/com/prescription/dto/PrescriptionResponse.java
//package com.prescription.dto;
//
//import com.fasterxml.jackson.annotation.JsonProperty;
//import lombok.Data;
//import lombok.NoArgsConstructor;
//import lombok.AllArgsConstructor;
//import lombok.Builder;
//
//import java.time.LocalDateTime;
//import java.util.List;
//
//@Data
//@NoArgsConstructor
//@AllArgsConstructor
//@Builder
//public class PrescriptionResponse {
//
//    @JsonProperty("id")
//    private Long prescriptionId;
//    private String patientName;
//    private String doctorName;
//    private LocalDateTime createdAt;
//    private String status;
//    private String imageUrl;
//
//    // ── Normalized medicines from Phase 3 ─────────────────────────────────
//    private List<MedicineResult> medicines;
//
//    // ── Phase 4a: Drug-Drug Interactions ──────────────────────────────────
//    private List<DrugInteractionResult> drugInteractions;
//
//    // ── Phase 4b: Drug→Disease warnings ──────────────────────────────────
//    private List<DrugDiseaseWarning> drugDiseaseWarnings;
//
//    // ── Summary counts for UI badges ──────────────────────────────────────
//    private int highSeverityCount;
//    private int moderateSeverityCount;
//
//    // =========================================================================
//    // Nested DTOs
//    // =========================================================================
//
//    @Data
//    @NoArgsConstructor
//    @AllArgsConstructor
//    @Builder
//    public static class MedicineResult {
//        private String originalName;
//        private String normalizedName;
//        private String rxcui;
//        private String medicineType;
//        private String dosage;
//        private String frequency;
//        private String duration;
//        private String quantity;
//        private double confidence;
//        private String method;     // exact_match | supplement_map | no_match | safety_blocked
//        private String status;     // completed | failed
//    }
//
//    @Data
//    @NoArgsConstructor
//    @AllArgsConstructor
//    @Builder
//    public static class DrugInteractionResult {
//        private String drug1Name;
//        private String drug1Rxcui;
//        private String drug2Name;
//        private String drug2Rxcui;
//        private String severity;      // HIGH | MODERATE | LOW | UNKNOWN
//        private String description;
//        private String source;        // DRUGBANK
//    }
//
//    @Data
//    @NoArgsConstructor
//    @AllArgsConstructor
//    @Builder
//    public static class DrugDiseaseWarning {
//        private String drugName;
//        private String rxcui;
//        private String indicationText;
//        private List<String> meshTerms;
//    }
//}


package com.prescription.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.Builder;

import java.time.LocalDateTime;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PrescriptionResponse {

    @JsonProperty("id")
    private Long prescriptionId;
    private String patientName;
    private String doctorName;
    private LocalDateTime createdAt;
    private String status;
    private String imageUrl;

    // ── Normalized medicines from Phase 3 ─────────────────────────────────
    private List<MedicineResult> medicines;

    // ── Phase 4a: Drug-Drug Interactions ──────────────────────────────────
    private List<DrugInteractionResult> drugInteractions;

    // ── Phase 4b: Drug→Disease warnings ───────────────────────────────────
    private List<DrugDiseaseWarning> drugDiseaseWarnings;

    // ── Duplicate active ingredient notices ───────────────────────────────
    @JsonProperty("duplicateIngredientWarnings")
    private List<DuplicateIngredientWarning> duplicateIngredientWarnings;

    // ── Summary counts for UI badges ──────────────────────────────────────
    private int highSeverityCount;
    private int moderateSeverityCount;

    // =========================================================================
    // Nested DTOs
    // =========================================================================

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class MedicineResult {
        private String originalName;
        private String normalizedName;
        private String rxcui;
        private String medicineType;
        private String dosage;
        private String frequency;
        private String duration;
        private String quantity;
        private double confidence;
        private String method;     // exact_match | supplement_map | no_match | safety_blocked
        private String status;     // completed | failed
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class DrugInteractionResult {
        private String drug1Name;
        private String drug1Rxcui;
        private String drug2Name;
        private String drug2Rxcui;
        private String severity;      // HIGH | MODERATE | LOW | UNKNOWN
        private String description;
        private String source;        // DRUGBANK
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class DrugDiseaseWarning {
        private String drugName;
        private String rxcui;
        private String indicationText;
        private List<String> meshTerms;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class DuplicateIngredientWarning {
        private String rxcui;
        private List<String> medicines;
        private String message;
    }
}

File: lib/models/prescription_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'prescription_response.g.dart';

/// Main API Response Model from FastAPI
@JsonSerializable()
class PrescriptionApiResponse {
  @JsonKey(name: 'request_id')
  final String requestId;

  @JsonKey(name: 'processing_time_sec')
  final double processingTimeSec;

  final PrescriptionResult result;

  PrescriptionApiResponse({
    required this.requestId,
    required this.processingTimeSec,
    required this.result,
  });

  factory PrescriptionApiResponse.fromJson(Map<String, dynamic> json) =>
      _$PrescriptionApiResponseFromJson(json);

  Map<String, dynamic> toJson() => _$PrescriptionApiResponseToJson(this);
}

/// Main Result object containing all prescription data
@JsonSerializable()
class PrescriptionResult {
  @JsonKey(name: 'patient_info')
  final Map<String, dynamic>? patientInfo;

  final Map<String, dynamic>? vitals;

  @JsonKey(name: 'clinical_info')
  final Map<String, dynamic>? clinicalInfo;

  final List<MedicineResult> prescription;

  @JsonKey(name: 'drug_interactions')
  final List<DrugInteractionResult> drugInteractions;

  @JsonKey(name: 'drug_disease_warnings')
  final List<DrugDiseaseWarning> drugDiseaseWarnings;

  @JsonKey(name: 'duplicate_ingredient_warnings')
  final List<DuplicateIngredientWarning> duplicateIngredientWarnings;

  @JsonKey(name: 'high_severity_count')
  final int highSeverityCount;

  @JsonKey(name: 'moderate_severity_count')
  final int moderateSeverityCount;

  final Map<String, dynamic>? audit;

  PrescriptionResult({
    required this.patientInfo,
    required this.vitals,
    required this.clinicalInfo,
    required this.prescription,
    required this.drugInteractions,
    required this.drugDiseaseWarnings,
    required this.duplicateIngredientWarnings,
    required this.highSeverityCount,
    required this.moderateSeverityCount,
    this.audit,
  });

  factory PrescriptionResult.fromJson(Map<String, dynamic> json) =>
      _$PrescriptionResultFromJson(json);

  Map<String, dynamic> toJson() => _$PrescriptionResultToJson(this);
}

/// Individual Medicine from prescription
@JsonSerializable()
class MedicineResult {
  @JsonKey(name: 'Medicine_Name')
  final String? medicineName;

  @JsonKey(name: 'Medicine_Type')
  final String? medicineType;

  @JsonKey(name: 'Dosage')
  final String? dosage;

  @JsonKey(name: 'Frequency')
  final String? frequency;

  @JsonKey(name: 'Duration_to_take_med')
  final String? duration;

  @JsonKey(name: 'Quantity')
  final String? quantity;

  final String? rxcui;

  @JsonKey(name: 'normalized_name')
  final String? normalizedName;

  @JsonKey(name: 'normalization_confidence')
  final double? normalizationConfidence;

  @JsonKey(name: 'normalization_status')
  final String? normalizationStatus;

  @JsonKey(name: 'normalization_method')
  final String? normalizationMethod;

  MedicineResult({
    required this.medicineName,
    this.medicineType,
    this.dosage,
    this.frequency,
    this.duration,
    this.quantity,
    this.rxcui,
    this.normalizedName,
    this.normalizationConfidence,
    this.normalizationStatus,
    this.normalizationMethod,
  });

  factory MedicineResult.fromJson(Map<String, dynamic> json) =>
      _$MedicineResultFromJson(json);

  Map<String, dynamic> toJson() => _$MedicineResultToJson(this);
}

/// Drug-Drug Interaction Warning
@JsonSerializable()
class DrugInteractionResult {
  @JsonKey(name: 'drug1_name')
  final String? drug1Name;

  @JsonKey(name: 'drug1_rxcui')
  final String? drug1Rxcui;

  @JsonKey(name: 'drug2_name')
  final String? drug2Name;

  @JsonKey(name: 'drug2_rxcui')
  final String? drug2Rxcui;

  final String severity; // HIGH | MODERATE | LOW | UNKNOWN

  final String? description;

  final String? source; // DRUGBANK

  DrugInteractionResult({
    required this.drug1Name,
    required this.drug1Rxcui,
    required this.drug2Name,
    required this.drug2Rxcui,
    required this.severity,
    this.description,
    this.source,
  });

  /// Returns color based on severity level
  String getSeverityColor() {
    switch (severity.toUpperCase()) {
      case 'HIGH':
        return '#d32f2f'; // Red
      case 'MODERATE':
        return '#f57c00'; // Orange
      case 'LOW':
        return '#fbc02d'; // Yellow
      default:
        return '#757575'; // Gray
    }
  }

  factory DrugInteractionResult.fromJson(Map<String, dynamic> json) =>
      _$DrugInteractionResultFromJson(json);

  Map<String, dynamic> toJson() => _$DrugInteractionResultToJson(this);
}

/// Drug-Disease Warning
@JsonSerializable()
class DrugDiseaseWarning {
  @JsonKey(name: 'drugName')
  final String? drugName;

  final String? rxcui;

  @JsonKey(name: 'indicationText')
  final String? indicationText;

  @JsonKey(name: 'meshTerms')
  final List<String>? meshTerms;

  DrugDiseaseWarning({
    required this.drugName,
    this.rxcui,
    this.indicationText,
    this.meshTerms,
  });

  factory DrugDiseaseWarning.fromJson(Map<String, dynamic> json) =>
      _$DrugDiseaseWarningFromJson(json);

  Map<String, dynamic> toJson() => _$DrugDiseaseWarningToJson(this);
}

/// Duplicate Active Ingredient Warning
@JsonSerializable()
class DuplicateIngredientWarning {
  final String? rxcui;

  final List<String> medicines;

  final String message;

  DuplicateIngredientWarning({
    this.rxcui,
    required this.medicines,
    required this.message,
  });

  factory DuplicateIngredientWarning.fromJson(Map<String, dynamic> json) =>
      _$DuplicateIngredientWarningFromJson(json);

  Map<String, dynamic> toJson() => _$DuplicateIngredientWarningToJson(this);
}

/// Local storage model for prescription history
@JsonSerializable()
class PrescriptionRecord {
  final String id;
  final String? imagePath;
  final DateTime uploadedAt;
  final PrescriptionApiResponse apiResponse;
  final String status; // processing, completed, failed

  PrescriptionRecord({
    required this.id,
    this.imagePath,
    required this.uploadedAt,
    required this.apiResponse,
    required this.status,
  });

  factory PrescriptionRecord.fromJson(Map<String, dynamic> json) =>
      _$PrescriptionRecordFromJson(json);

  Map<String, dynamic> toJson() => _$PrescriptionRecordToJson(this);
}
